### PR TITLE
Improve retry mechanism

### PR DESCRIPTION
**Pull request by pablopalacios**
@redonkulus while thinking in a way of adding retry mechanism to fetchr to solve an issue in my company, I discovered that fetchr already has support for it. I went deeper on it and realized that there was a bug with the retry interval and that the status code check was only supporting 408 and xhr failures. Then I fixed the bug and add support for custom status while keeping it backward compatible.

- fix initial retry.interval not being respected
- add support for custom response status codes to be retried automatically

**UPDATE:**
- add full jitter formula for interval calculation
- deprecate `max_retries` in favor of `maxRetries` 
- add documentation

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: libs/util/http.client.js [hunk 1]
```diff
@@ -103,41 +107,34 @@ function mergeConfig(config) {
...
+            if (config.retry.statusCodes) {
+                cfg.retry.statusCodes = config.retry.statusCodes
+            }
         }
 
         if (config.withCredentials) {
             cfg.withCredentials = config.withCredentials;
         }
-
-        // tmp stores transient state data, such as retry count
```
Comment (redonkulus) [inline libs/util/http.client.js:120]: Why was this removed?

File: libs/util/http.client.js [hunk 2]
```diff
@@ -103,41 +107,34 @@ function mergeConfig(config) {
...
             if (!isNaN(interval) && interval > 0) {
                 cfg.retry.interval = interval;
             }
-            maxRetries = parseInt(config.retry && config.retry.max_retries, 10);
+            maxRetries = parseInt(config.retry.max_retries, 10);
             if (!isNaN(maxRetries) && maxRetries >= 0) {
                 cfg.retry.max_retries = maxRetries;
             }
+            if (config.retry.statusCodes) {
+                cfg.retry.statusCodes = config.retry.statusCodes
```
Comment (redonkulus) [inline libs/util/http.client.js:119]: Lets document this new option in the readme please.

**Review by redonkulus — CHANGES_REQUESTED** [review]
Thanks for the PR!

Looks like I need to run prettier on this code base, its showing its age. Left a couple of comments. Also tagged @lingyan to take a look since she might have more knowledge of this retry logic than me.

File: libs/util/http.client.js (hunk 1 above)
Comment (pablopalacios) [reply libs/util/http.client.js:120]: **Context**: for every retry, we double the interval for the next request. And, if the amount of retries reaches the maximum limit, `fetchr` just give up and returns an error.

**Previous implementation:** `cfg.tmp` was only being used to store the number of attempts and the interval for the next request. Whenever we would call `shouldRetry`, we would mutate this object by increasing the number of attempts and by doubling the interval so the next retry could be properly set.

 What I did was to replace this mutable object with an explicit required `attempt` param in `doXhr` function. `doXhr` is the function that is responsible for attempting the request, calling `shouldRetry` and scheduling a new request attempt if necessary (by calling itself). 

Since `doXhr` has `attempt` as param now, it can just pass `attempt + 1` to the inner `doXhr` call. Also, we can get the new interval value by just doing `Math.pow(2, attempt) * interval`. So no need for a mutable object that is shared between calls.

**Comment by pablopalacios** [comment]
> Looks like I need to run prettier on this code base, its showing its age. 

That would be so awesome!

**Comment by redonkulus** [comment]
Added here, this will mess up your PR once merged https://github.com/yahoo/fetchr/pull/261

File: libs/util/http.client.js [hunk 3]
```diff
@@ -146,12 +143,14 @@ function doXhr(method, url, headers, data, config, callback) {
...
+                if (!shouldRetry(method, config, response.statusCode, attempt)) {
                     callback(err);
                 } else {
                     setTimeout(
-                        function retryXHR() { doXhr(method, url, headers, data, config, callback); },
-                        config.retry.interval
+                        function retryXHR() {
+                            doXhr(method, url, headers, data, config, attempt + 1, callback);
+                        },
+                        config.retry.interval * Math.pow(2, attempt)
```
Comment (lingyan) [inline libs/util/http.client.js:153]: @pablopalacios since you are at this, do you mind improving this exponential retry logic to use full jitter, as documented [here](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) by aws?

**Review by lingyan — COMMENTED** [review]
I like the changes! Logic is cleaner.  Left a comment for enhancement.  Thanks!

File: libs/util/http.client.js (hunk 3 above)
Comment (pablopalacios) [reply libs/util/http.client.js:153]: What should we use as cap? Or should we not use cap at all?

File: libs/util/http.client.js (hunk 2 above)
Comment (pablopalacios) [reply libs/util/http.client.js]: Done. But while documenting I felt that it was so wrong to use `max_retries` while all the other settings use camel case names that I added another commit deprecating `max_retries` in favor of `maxRetries`.

File: libs/util/http.client.js (hunk 3 above)
Comment (pablopalacios) [reply libs/util/http.client.js]: @lingyan I updated it but I haven't set `cap` for now so the delay will always be a random number between 0 and `                        config.retry.interval * Math.pow(2, attempt)`. My feeling was that setting a `cap` as in the aws article would require us to set another property in the configuration to allow users to tweak it properly if they have different interval configurations. We could instead, in the future, add another configuration where users can pass a custom function to calculate the delay.

Comment (lingyan) [reply libs/util/http.client.js]: Sure!  I am ok with not using cap.  We can add it later if needed.  Thanks @pablopalacios for making the change!

File: libs/util/http.client.js [hunk 4]
```diff
@@ -152,12 +155,27 @@ function doXhr(method, url, headers, data, config, callback) {
                 callback(NULL, response);
             },
             failure: function (err, response) {
-                if (!shouldRetry(method, config, response.statusCode)) {
+                if (
+                    !shouldRetry(method, config, response.statusCode, attempt)
+                ) {
                     callback(err);
                 } else {
+                    var delay =
```
Comment (lingyan) [inline libs/util/http.client.js:163]:
```suggestion
                    // Use exponential backoff and full jitter strategy published in https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
                    var delay =
```

**Review by lingyan — COMMENTED** [review]
Left two doc related suggestions. minor. rest looks good. Thanks!

**Review by lingyan — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -430,6 +430,69 @@ fetcher.read('service').params({ id: 1 }).clientConfig(config).end(callbackFn);
...
+            maxRetries: 2,
+        },
+    })
+    .end();
+```
+
+With this confiugration, Fetchr will retry all requests that fails
+with 408 status code or with a XHR 0 status code for 2 more times
+before returning an error. The interval between each requests respects
+the following formula, based on the exponential backoff and full jitter strategy published in [this AWS architecture blog post](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/):
```
Comment (redonkulus) [inline README.md:452]:
```suggestion
With this configuration, Fetchr will retry all requests that fail with 408 status code or with an XHR 0 status code two more times before returning an error. The interval between each request respects
the following formula, based on the exponential backoff and full jitter strategy published in [this AWS architecture blog post](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/):
```

File: README.md [hunk 6]
```diff
@@ -430,6 +430,67 @@ fetcher.read('service').params({ id: 1 }).clientConfig(config).end(callbackFn);
...
+```
+
+With the above configuration, Fetchr will retry for at maxiumum of 5
+times all the requests that fails with a 408 or 502 status code. The
+interval between each request will still use the formula from above,
+but the interval of 1000ms will be used instead.
+
+**Note:** Fetchr doesn't retry POST requests for safety reasons. You
+can enable retries for POST requests by setting the `unsafeAllowRetry`
+property to `true`:
```
Comment (redonkulus) [inline README.md:482]:
```suggestion
With the above configuration, Fetchr will retry all failed (408 or 502 status code) requests for a maximum of 5 times. The interval between each request will still use the formula from above, but the interval of 1000ms will be used instead.

**Note:** Fetchr doesn't retry POST requests for safety reasons. You can enable retries for POST requests by setting the `unsafeAllowRetry` property to `true`:
```

File: README.md [hunk 7]
```diff
@@ -430,6 +430,62 @@ fetcher.read('service').params({ id: 1 }).clientConfig(config).end(callbackFn);
 
 For requests from the server, the config object is simply passed into the service being called.
 
+## Retry
+
+You can set Fetchr to retry failed requests automatically by setting a
+`retry` settings in the client configuration:
```
Comment (redonkulus) [inline README.md:436]:
```suggestion
You can set Fetchr to retry failed requests automatically by setting a `retry` settings in the client configuration:
```

**Review by redonkulus — APPROVED** [review]

**Comment by redonkulus** [comment]
Published https://github.com/yahoo/fetchr/releases/tag/v0.5.45
